pkg/kubelet/kuberuntime/kuberuntime_manager_test.go

```diff
@@ -1230,14 +1357,289 @@ func TestComputePodActions(t *testing.T) {
+			},
+		},
+		"pod marked for RestartAllContainers with init containers": {
+			podFunc: func() *v1.Pod {
+				pod, _ := makeBasePodAndStatusWithInitContainers()
+				pod.Status.Conditions = allContainersRestartingTrue
+				return pod
+			},
+			podStatusFunc: func() *kubecontainer.PodStatus {
+				_, status := makeBasePodAndStatusWithInitContainers()
+				return status
+			},
+			containersToRemove: []containerToRemoveInfo{
+				{name: "init3"},
+				{name: "init2"},
+				{name: "init1"},
+			},
+		},
+		"pod marked for RestartAllContainers with restartable init containers": {
+			podFunc: func() *v1.Pod {
+				pod, _ := makeBasePodAndStatusWithRestartableInitContainers()
+				pod.Status.Conditions = allContainersRestartingTrue
+				return pod
+			},
+			podStatusFunc: func() *kubecontainer.PodStatus {
+				_, status := makeBasePodAndStatusWithRestartableInitContainers()
+				return status
+			},
+			containersToRemove: []containerToRemoveInfo{
+				{name: "restartable-init-3", kill: true},
+				{name: "restartable-init-2", kill: true},
+				{name: "restartable-init-1", kill: true},
+			},
+		},
+		"init container exit triggers RestartAllContainres": {
+			// The init3 container exited and triggers RestartAllContainers,
+			// the init2 container is a running sidecar. First, the init2
+			// should be killed and removed; second, the init1 containers should
+			// be removed; lastly, init3 container should be removed.
+			podFunc: func() *v1.Pod {
+				pod, _ := makeBasePodAndStatusWithInitContainers()
+				sidecar := pod.Spec.InitContainers[1]
+				sidecar.RestartPolicy = &restartPolicyAlways
+				pod.Spec.InitContainers[1] = sidecar
+				source := pod.Spec.InitContainers[2]
+				source.RestartPolicy = &restartPolicyNever
+				source.RestartPolicyRules = restartAllContainersRules
+				pod.Spec.InitContainers[2] = source
+				pod.Status.Conditions = allContainersRestartingTrue
+				return pod
+			},
+			podStatusFunc: func() *kubecontainer.PodStatus {
+				_, status := makeBasePodAndStatusWithInitContainers()
+				sidecarStatus := status.ContainerStatuses[1]
+				sidecarStatus.State = kubecontainer.ContainerStateRunning
+				status.ContainerStatuses[1] = sidecarStatus
+				sourceStatus := status.ContainerStatuses[2]
+				sourceStatus.ExitCode = 1
+				status.ContainerStatuses[2] = sourceStatus
+				return status
+			},
+			containersToRemove: []containerToRemoveInfo{
+				{name: "init2", kill: true},
+				{name: "init1"},
+				{name: "init3"},
+			},
+		},
+		"sidecar container exit triggers RestartAllContainres": {
+			// Restartable-init-3 fails and triggers RestartAllContainers.
+			// The running foo1 should be killed and removed first; then init-2 and init-1
+			// should be killed and removed; lastly init-3 should be removed.
+			podFunc: func() *v1.Pod {
+				pod, _ := makeBasePodAndStatusWithRestartableInitContainers()
+				source := pod.Spec.InitContainers[2]
+				source.RestartPolicy = &restartPolicyNever
```

Comment:
Since `init-3` is a restartable init container, its `restartPolicy` should remain Always.

```suggestion

```